trino/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergHiveMetastoreAutoCleanMetadataFile.java

```diff
@@ -0,0 +1,170 @@
+                                .put("s3.max-connections", "2") // verify no leaks
+                                .put("iceberg.register-table-procedure.enabled", "true")
+                                .put("iceberg.writer-sort-buffer-size", "1MB")
+                                .buildOrThrow())
+                .setSchemaInitializer(
+                        SchemaInitializer.builder()
+                                .withSchemaName("tpch")
+                                .withSchemaProperties(Map.of("location", "'s3://" + bucketName + "/tpch'"))
+                                .build())
+                .build();
+        HiveMetastore metastore = ((IcebergConnector) queryRunner.getCoordinator().getConnector(ICEBERG_CATALOG)).getInjector()
+                .getInstance(HiveMetastoreFactory.class)
+                .createMetastore(Optional.empty());
+
+        this.fileSystemFactory = getFileSystemFactory(queryRunner);
+        this.tableOperationsProvider = new FileMetastoreTableOperationsProvider(fileSystemFactory);
+
+        CachingHiveMetastore cachingHiveMetastore = createPerTransactionCache(metastore, 1000);
+        this.trinoCatalog = new TrinoHiveCatalog(
+                new CatalogName("hive"),
+                cachingHiveMetastore,
+                new TrinoViewHiveMetastore(cachingHiveMetastore, false, "trino-version", "test"),
+                fileSystemFactory,
+                new TestingTypeManager(),
+                tableOperationsProvider,
+                false,
+                false,
+                false,
+                new IcebergConfig().isHideMaterializedViewStorageTable());
+
+        return queryRunner;
+    }
+
+    @Test
+    public void testInsertWithAutoCleanMetadataFile()
+            throws IOException
+    {
+        assertUpdate("CREATE TABLE table_to_metadata_count (_bigint BIGINT, _varchar VARCHAR)");
+
+        Table table = IcebergUtil.loadIcebergTable(trinoCatalog, tableOperationsProvider, TestingConnectorSession.SESSION,
+                new SchemaTableName("tpch", "table_to_metadata_count"));
+        table.updateProperties()
+                .set(TableProperties.METADATA_DELETE_AFTER_COMMIT_ENABLED, "true")
+                .set(TableProperties.METADATA_PREVIOUS_VERSIONS_MAX, String.valueOf(METADATA_PREVIOUS_VERSIONS_MAX))
+                .commit();
+        for (int i = 0; i < 10; i++) {
+            assertUpdate("INSERT INTO table_to_metadata_count VALUES (1, 'a')", 1);
+        }
+        int count = 0;
+        fileSystem = fileSystemFactory.create(SESSION);
+        FileIterator fileIterator = fileSystem.listFiles(Location.of(table.location()));
+        while (fileIterator.hasNext()) {
+            FileEntry next = fileIterator.next();
+            if (next.location().path().endsWith("metadata.json")) {
+                count++;
+            }
+        }
+        assertThat(count).isEqualTo(1 + METADATA_PREVIOUS_VERSIONS_MAX);
```

Comment:
We should verify that the oldest metadata file was exactly removed instead of just checking the count.